src/celeritas/geo/GeoMaterialParams.cc

```diff
@@ -28,6 +28,211 @@
+    // build a map of volume name -> matid
+    for (auto idx : range(materials.size()))
+    {
+        if (!materials[idx])
+        {
+            // Skip volumes without matids
+            continue;
+        }
+
+        auto [prev, inserted]
+            = lab_to_id.insert({std::move(labels[idx]), materials[idx]});
+        if (!inserted)
+        {
+            duplicates.insert(prev->first);
+        }
+    }
+    CELER_VALIDATE(duplicates.empty(),
+                   << "geo/material coupling specified duplicate "
+                      "volume names: \""
+                   << join(duplicates.begin(), duplicates.end(), "\", \"")
+                   << '"');
+    return lab_to_id;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find a material ID from a volume ID.
+ */
+class MaterialFinder
+{
+  public:
+    MaterialFinder(GeoParams const& geo, MapLabelMatId const& materials)
+        : geo_{geo}, materials_{materials}
+    {
+    }
+
+    MaterialId operator()(VolumeId const& volume_id)
+    {
+        Label const& vol_label = geo_.id_to_label(volume_id);
+
+        // Hopefully user-provided and geo-provided volume labels match exactly
+        if (auto iter = materials_.find(vol_label); iter != materials_.end())
+        {
+            return iter->second;
+        }
+
+        if (mat_labels_.empty())
+        {
+            // Build user volume -> material mapping
+            this->build_mat_labels();
+        }
+
+        // Either:
+        // - user-provided vol labels have no extensions? (just names)
+        // - geometry volume labels are missing extensions (e.g. when
+        //   Geant4-derived volume names, ORANGE geometry names)
+        auto [start, stop] = mat_labels_.equal_range(vol_label.name);
+        if (start == stop)
+        {
+            // No materials match the volume label
+            return {};
+        }
+        if (std::distance(start, stop) == 1)
+        {
+            return start->second.second;
+        }
+
+        // Multiple labels match
+        CELER_LOG(warning)
+            << "Multiple materials match the volume '" << vol_label << "': "
+            << join_stream(
+                   start, stop, ", ", [](std::ostream& os, auto&& mliter) {
+                       os << mliter.second.second.unchecked_get();
+                   });
+        return start->second.second;
+    }
+
+  private:
+    GeoParams const& geo_;
+    MapLabelMatId const& materials_;
+
+    using PairExtMatid = std::pair<std::string, MaterialId>;
+    std::multimap<std::string, PairExtMatid> mat_labels_;
+
+    void build_mat_labels()
+    {
+        for (auto const& [mlabel, matid] : materials_)
+        {
+            mat_labels_.emplace(mlabel.name, PairExtMatid{mlabel.ext, matid});
+        }
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Whether a volume with a missing material needs to be reported to the user.
```

Comment:
Should the documentation describe and/or refers to the documentation of the heuristic used.